### PR TITLE
Refactor workspace instance `attributedTeamId` to an explicit, not-team-specific `usageAttributionId`

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Refactor workspace instance `attributedTeamId` to an explicit, not-team-specific `usageAttributionId`.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/10822

## How to test

1. Should build
2. Workspace instances should always get a `usageAttributionId`
    - If the workspace is for a project that's owned by a Team, the attribution ID should look like `team:1234`
    - If the workspace is for a user project, or for no project at all, the attribution ID should look like `user:1234`

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:14]: FYI, I tried `ALGORITHM=INPLACE`, but MySQL wasn't happy about it:

```
query failed: ALTER TABLE `d_b_workspace_instance` CHANGE `attributedTeamId` `usageAttributionId` varchar(255) NOT NULL DEFAULT '', ALGORITHM=INPLACE, LOCK=NONE
error: Error: ER_ALTER_OPERATION_NOT_SUPPORTED_REASON: ALGORITHM=INPLACE is not supported. Reason: Cannot change column type INPLACE. Try ALGORITHM=COPY.
    at Query.Sequence._packetToError (/tmp/build/components-gitpod-db--dbtest-init.2db8f1d34fe1fa534d23cb857822e9eb7e0f3398/mig/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
```

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:14]: Oh no, `LOCK=NONE` also doesn't work 🙈

```
error: Error: ER_ALTER_OPERATION_NOT_SUPPORTED_REASON: LOCK=NONE is not supported. Reason: Cannot change column type INPLACE. Try LOCK=SHARED.
```

**Comment by jankeromnes** [comment]
It's working! 🎉

```
mysql> select id, workspaceId, creationTime, usageAttributionId from d_b_workspace_instance;
+--------------------------------------+-----------------------------------+--------------------------+-------------------------------------------+
| id                                   | workspaceId                       | creationTime             | usageAttributionId                        |
+--------------------------------------+-----------------------------------+--------------------------+-------------------------------------------+
| 3610efe0-28f3-43f9-a742-62cf2b15c361 | autofixdev-autofix-qdb4hnp3343    | 2022-06-23T15:01:14.804Z | user:03c5bb79-361d-4956-ad0e-58221b378484 |
| 7e62f39b-8e70-40ea-9fe3-6776e3d51c09 | jankeromnes-metachess-ne9ernqjojg | 2022-06-23T15:02:12.145Z | team:b2ee4f72-2dab-48b9-9b2b-903387e577f2 |
| f63a6dcd-04ca-4ca7-b4ca-d2bdf5ca70af | jankeromnes-metachess-u8zwfhjtiqq | 2022-06-23T15:01:59.035Z | team:b2ee4f72-2dab-48b9-9b2b-903387e577f2 |
+--------------------------------------+-----------------------------------+--------------------------+-------------------------------------------+
3 rows in set (0.01 sec)
```

File: components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: > varchar(255)

What about making it smaller in max size? Mentioning this because the bigger the key, the bigger the index and the more costly operations on it will be. :watch:

File: components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:25]: :+1:

File: components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: Good catch! But `varchar`'s exact size depends on the actual content, right? And with one address byte, you can go up to 255 characters. So, I think there is no advantage in using `varchar(50)` over `varchar(255)`, other than preventing ourselves from using larger sizes one day (either by accident or because we need to).

From the [Drupal docs](https://www.drupal.org/docs/7/api/schema-api/data-types/varchar)

> **Storage Information :**
> L + 1 bytes if column values require 0 − 255 bytes, L + 2 bytes if values may require more than 255 bytes where L is L is the byte length of the string that you are storing and not the maximum size that you have set.

Comment (geropl) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: That's true for the column, but not for the indexex: It has to assume the max size for a given column. And that directly affects query times, especially if the index becomes too big for RAM (as we see for this table already).

If we want to later `GROUP BY` on this, an index is required.

**Comment by geropl** [comment]
Nice.
image

File: components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: Ah, that's an excellent point, thank you @geropl!

So, should I:
- change the type to `char(41)` (exactly enough for `team:` or `user:` prefix followed by UUID)?
- or make it slightly bigger? (e.g. `char(50)`?)
- and add an index to it in this migration?

Comment (geropl) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: I'm fine with sth like `char(50)` or `char(60)` to anticipate slighty longer prefixes. :+1: 

> and add an index to it in this migration?

Let's leave that for a follow-up. :lotus_position:

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/migration/1655988518555-WorkspaceInstanceUsageAttributionId.ts:17]: > I'm fine with sth like `char(50)` or `char(60)` to anticipate slighty longer prefixes. 👍

How about `varchar(60)`? 😁

> Let's leave that for a follow-up. 🧘

I think it's just a couple more characters in the migration, right? 😇 Maybe I can add it now.

**Review by geropl — APPROVED** [review]
Tested and works, approving to unblock.

/hold So for [reducing the column size](https://github.com/gitpod-io/gitpod/pull/10868/files#r905842410). :+1:

**Comment by jankeromnes** [comment]
@geropl FYI, I've changed the type to `varchar(60)` and created an index: https://github.com/gitpod-io/gitpod/compare/303f5aa3e8ee03eeb22724b355e608e349d1fe1c..b3c03b94c67f550292645717393a112af1f8c054

**Comment by jankeromnes** [comment]
Still working as intended! 🎉 Many thanks for the review @geropl 🙏

/unhold

**Pull request merged**
